File: MobileMeteo/MMTMeteorogramFetchDelegate.m
//
//  MMTMeteorogramFetchDelegate.m
//  MobileMeteo
//
//  Created by Kamil Szostakowski on 08.06.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

#import "MMTMeteorogramFetchDelegate.h"

@interface MMTMeteorogramFetchDelegate ()

@property (nonatomic, strong) NSData *responseData;
@property (nonatomic, strong) MMTMeteorogramFetchCompletion completion;
@end

@implementation MMTMeteorogramFetchDelegate

- (instancetype)initWithCompletion:(MMTMeteorogramFetchCompletion)completion
{
    if(completion)
    {
        self = [super init];
        
        if(self)
            _completion = completion;
        
        return self;
    }
    
    return nil;
}

#pragma mark - Delegate methods

- (NSURLRequest *)connection:(NSURLConnection *)connection willSendRequest:(NSURLRequest *)request redirectResponse:(NSURLResponse *)response
{
    if(response)
    {
        NSHTTPURLResponse *redirectResponse = (NSHTTPURLResponse *)response;
        NSString *locationUri = redirectResponse.allHeaderFields[@"Location"];
    
        NSString *queryString = [[locationUri componentsSeparatedByString:@"?"] lastObject];
        queryString = [NSString stringWithFormat:@"?%@", queryString];
    
        NSURL *redirectUrl = [NSURL URLWithString:queryString relativeToURL:[NSURL mmt_meteorogramDownloadBaseUrl]];

        return [NSURLRequest requestWithURL:redirectUrl];
    }
    
    return request;
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
{
    NSMutableData *responseData = self.responseData ? self.responseData.mutableCopy : data;
    
    if (self.responseData)
    {
        [responseData appendData:data];
    }
            
    self.responseData = [NSData dataWithData:responseData];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection
{
    self.completion(self.responseData, nil);
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error
{
    self.responseData = nil;
    self.completion(nil, error);
}

@end

File: MobileMeteo/MMTMeteorogramFetchDelegate.swift
//
//  MMTMeteorogramFetchDelegate.swift
//  MobileMeteo
//
//  Created by Kamil Szostakowski on 20.08.2015.
//  Copyright (c) 2015 Kamil Szostakowski. All rights reserved.
//

import Foundation

public typealias MMTFetchMeteorogramCompletion = (data: NSData?, error: NSError?) -> Void

public class MMTMeteorogramFetchDelegate: NSObject, NSURLConnectionDataDelegate
{
    // MARK: Properties
    
    private let MMTEmptyMeteorogramSize = 71
    
    let completion: MMTFetchMeteorogramCompletion
    var responseData: NSData?
    
    // MARK: Initializers
    
    public init(completion: MMTFetchMeteorogramCompletion)
    {
        self.completion = completion;
        
        super.init()
    }
    
    // MARK: Protocol methods
    
    public func connection(connection: NSURLConnection, didReceiveData data: NSData)
    {
        let responseData = self.responseData != nil ?
            NSMutableData(data: self.responseData!) :
            NSMutableData(data: data)
        
        if self.responseData != nil {
            responseData.appendData(data)
        }
        
        self.responseData = NSData(data: responseData)
    }
    
    public func connectionDidFinishLoading(connection: NSURLConnection)
    {
        switch responseData?.length <= MMTEmptyMeteorogramSize
        {
            case true: completion(data: nil, error: errorWithCode(.MeteorogramNotFound))
            case false: completion(data: responseData, error: nil)
            
            default: break
        }
    }
    
    public func connection(connection: NSURLConnection, didFailWithError error: NSError)
    {
        responseData = nil
        completion(data: nil, error: errorWithCode(.MeteorogramFetchFailure))
    }
    
    // MARK: Helper methods
    
    private func errorWithCode(code: MMTError) -> NSError
    {
        return NSError(domain: MMTErrorDomain, code: code.rawValue, userInfo: nil)
    }
}